### ISSUE TITLE
NGLparsererror

### ISSUE THREAD

**Issue by pr315** [nvuillam/npm-groovy-lint#226]
Hey i am using the npm-groovy-lint on my travis config file to lint only jenkins files in my repo and there is one error which is not getting fixed, can you please help me out.

Error Message : 6     error    Unexpected input: '{' @ line 6, column 10.  NglParseError 

Jenkins File : 

pipeline {

    agent {
         label 'custom-cp4s'
    }
    environment {
        CLUSTER_HOST = "icp-console.apps.coherent-orca-icp-mst.fyre.abc.com"
        REPO_FOUNDATIONS = "git@github.abc.com:security-secops/abc-security-foundations-prod.git"
        MISSING_REPOS = "abc-security-foundations-prod abc-security-solutions-prod"
        MISSING_ALIAS = "foundations solutions"
    }
    parameters {
        string(name: 'CHART', description: 'The chart to be packaged. Full chart name or alias can be used, Eg tis or abc-isc-tis-prod. For all charts leave as default value', defaultValue: 'All')
        string(name: 'CHART_BRANCH', description: 'The chart repository branch to be updated. e.g master,develop,release/1.5.0.0', defaultValue: '')
        string(name: 'FOUNDATIONS_BRANCH', description: 'The Foundations chart repository branch to be updated only required when updating abc-security-solutions-prod. e.g master,develop,release/1.5.0.0', defaultValue: '')
        choice(name: 'APP_VERSION',description: 'CP4S app version, i.e 1.4.0.1, 1.5.0.0', choices: ['1.5.0-0','1.5.0-1','1.5.0-2','1.5.0-3','1.5.0-4','1.5.0-5','1.6.0-0','1.6.0-1','1.6.0-2','1.6.0-3','1.6.0-4','1.6.0-5','1.7.1-0','1.7.1-1','1.7.2-0','1.7.2.1','1.7.2-2','1.7.2-3','1.7.2-4','1.7.2-5','1.8.0-0','1.8.0-1','1.9.0-0'])
        choice(name: 'CHART_VERSION',description: 'Chart/Case version number', choices: ['1.0.16','1.0.17','1.0.18','1.0.19','1.0.20','1.0.21','1.0.21','1.0.22','1.0.23','1.0.24','1.0.25',,'1.0.26','1.0.27'])
        string(name: 'RELEASE_NOTES', description: 'Details of the cp4s release to be added in release notes (Optional)', defaultValue: '')
    }
    stages {
        stage("tools_and_validation") {
            steps {
                    withCredentials([string(credentialsId: 'GITHHUB_TOKEN', variable: 'GITHUB_TOKEN')])
                    {
                    sh '''#!/bin/bash
            ### Fetching required binaries ###
            source ./scripts/fetch_binaries2.sh
            binaries::installCloudctl
            binaries::installOC
            binaries::install_casectl $GITHUB_TOKEN

                    '''
                    }
            }
        }

        // gets list of charts from solutions charts.yml
        stage("get charts") {
            steps {
                sshagent(['acbeeca1-f1f8-4021-bb58-0e36b7f81757']) {
                    sh '''#!/bin/bash
                        chmod +x ./scripts/get_charts_list.sh
                        CHARTS=$(./scripts/get_charts_list.sh "${CHART}" "${MISSING_REPOS}" "${MISSING_ALIAS}")

                        if [ $? -ne 0  ]; then
                            echo "$CHARTS"
                            exit 1
                        fi

                        echo "-------"
                        echo $CHARTS > charts_list.txt
                        cat charts_list.txt
                        echo "-------"
                    '''
                }

                // set CHART as global env var so it can be used in later stages
                script {
                  CHART = sh(returnStdout: true, script: 'cat charts_list.txt |tr -d "\n"')
                  env.CHART = "$CHART"
                }
            }
        }

        stage("scm") {
            steps {
                sshagent(['acbeeca1-f1f8-4021-bb58-0e36b7f81757']) {
                    sh '''#!/bin/bash

                        for var in ${CHART[@]}
                        do
                            REPO="git@github.abc.com:security-secops/${var}.git"

                            if [[ ${CHART} == "abc-security-solutions-prod" ]]; then
                                 git clone ${REPO_FOUNDATIONS} -b ${FOUNDATIONS_BRANCH}
                            fi

                            if ! git clone $REPO -b ${CHART_BRANCH}; then
                                echo "failed to clone repo"
                                exit 1
                            fi
                        done
                    '''
                }
            }
        }

        stage("Update Chart and Case"){
          steps{
              sshagent(['acbeeca1-f1f8-4021-bb58-0e36b7f81757']) {
                  sh'''#!/bin/bash

                      for var in ${CHART[@]}
                      do
                          if ! bash ./scripts/update_release.sh "$var" "${CHART_VERSION}" "${APP_VERSION}" "${RELEASE_NOTES}"; then
                              exit 1
                          fi

                          cd $var
                          git config --global push.default simple
                          git add ./stable/
                          git commit -m " Update $var chart/case version"
                          git push origin ${CHART_BRANCH}
                          cd ..

                          if [[ ${var} == "abc-security-solutions-prod" ]]; then
                             cd abc-security-foundations-prod
                             git config --global push.default simple
                             git add ./stable/
                             git commit -m " Update $var chart/case version"
                             git push origin ${FOUNDATIONS_BRANCH}
                             cd ..
                          fi
                      done
                  '''
              }
          }
        }
    }
}

**Comment by nvuillam** [comment]
You have a double comma in one of your choice parameters :)

![image](https://user-images.githubusercontent.com/17500430/183307705-ab4c7483-1bad-40eb-8686-b7b9128b0db9.png)

**Comment by pr315** [comment]
Thank you so much !! that helped. Your Top Notch!! :)

**Comment by nvuillam** [comment]
you're welcome :)

**Comment by pr315** [comment]
Hey @nvuillam , can you please help me out with this Jenkins File as well, I am really sorry for asking again.

Error : 52    error    Unexpected input: '{' @ line 52, column 10.  NglParseError

Jenkins File : 

def loadCommonVars() {

    vault_key_name = 'bootnode_privatekey_' + AWS_ACCOUNT + '_' + REGION

    if (AWS_ACCOUNT.contains('dev')) {
        credentials_location = 'secret/CP4SaaS/AWS/credentials'
        credID = 'vault-cred-id'
    } else {
        credentials_location = 'secret/CP4SaaS/production/AWS/credentials'
        credID = 'cp4saas_vault_env_prod'
    }

    vault_config = withCredentials([string(credentialsId: 'SAAS_VAULT_URL', variable: 'saas_vault_url')]) { [
        vaultUrl: saas_vault_url,
        skipSslVerification: 'true',
        vaultCredentialId: credID,
        engineVersion: 2
    ]}

    if (AWS_ACCOUNT.equals('dev_infra')) {
        vault_secrets = [
        [path: credentials_location, engineVersion: 2, secretValues: [
          [ envVar: 'AWS_ACCESS_KEY_ID', vaultKey: params.AWS_ACCOUNT + '_aws_access_key_id' ],
          [ envVar: 'AWS_SECRET_ACCESS_KEY', vaultKey: params.AWS_ACCOUNT + '_aws_secret_access_key' ]]
        ]
      ]
    } else {
        vault_secrets = [
        [path: credentials_location, engineVersion: 2, secretValues: [
          [ envVar: 'AWS_ADMIN_ROLE_ARN', vaultKey: params.AWS_ACCOUNT + '_aws_admin_role_arn' ]]
        ]
      ]
    }
}

node {
    script {
        // building the account choices out of env type
        wks = "${env.WORKSPACE}"
        if ( wks =~ /icp\/cp4saas\/protected-jobs\//) {
            ACCOUNT_CHOICE = [ 'production_uip' ]
        } else {
            ACCOUNT_CHOICE = [ 'dev_infra', 'dev_uip' ]
        }
    }
}

pipeline {
    agent any

    parameters {
        string(name: 'REGION', description: 'Insert the Stack Region', defaultValue: '')
        choice(name: 'AWS_ACCOUNT', description: 'Select Environment Type', choices: ACCOUNT_CHOICE)
        password(name: 'AWS_ACCESS_KEY_ID', description: 'Required when selected AWS Account is dev_uip or production_uip')
        password(name: 'AWS_SECRET_ACCESS_KEY', description: 'Required when selected AWS Account is dev_uip or production_uip')
        password(name: 'AWS_SESSION_TOKEN', description: 'Required when selected AWS Account is dev_uip or production_uip')
    }

    stages {
        stage('CLI tools') {
            steps {
                script {
                    sh'''#!/bin/bash
                        set -e
                        echo "Installing AWS CLI"
                        curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip"
                        unzip awscliv2.zip > /dev/null
                        sudo ./aws/install
                        aws --version
                    '''
                }
            }
        }
        stage('Generate Key') {
            steps {
                script {
                    currentBuild.displayName += " Env: $AWS_ACCOUNT - Region: $REGION"

                    loadCommonVars()

                    withVault([configuration: vault_config, vaultSecrets: vault_secrets]) {
                        script {
                            // if uip account, use supplied mfa creds as env vars for aws cli
                            if (params.AWS_ACCOUNT != 'dev_infra') {
                                env.AWS_ACCESS_KEY_ID = "${params.AWS_ACCESS_KEY_ID}"
                                env.AWS_SECRET_ACCESS_KEY = "${params.AWS_SECRET_ACCESS_KEY}"
                                env.AWS_SESSION_TOKEN = "${params.AWS_SESSION_TOKEN}"
                                sh'''
                              #!/bin/bash
                              set -xe
                              aws sts assume-role --role-arn "$AWS_ADMIN_ROLE_ARN" --role-session-name "admin@cluster-provisioning" > admin_role.json
                            '''
                                env.AWS_ACCESS_KEY_ID = sh(returnStdout: true, script: 'jq -rj ".Credentials.AccessKeyId" admin_role.json')
                                env.AWS_SECRET_ACCESS_KEY = sh(returnStdout: true, script: 'jq -rj ".Credentials.SecretAccessKey" admin_role.json')
                                env.AWS_SESSION_TOKEN = sh(returnStdout: true, script: 'jq -rj ".Credentials.SessionToken" admin_role.json')
                            }
                        }
                        sh'''
                            #!/bin/bash
                            set +x -e
                            KEY_NAME="xdr-bootnode"

                            if [ -z "$REGION" ]; then
                                echo "REGION is not set"
                                exit 1
                            fi

                            aws configure set region $REGION

                            aws ec2 create-key-pair --key-name $KEY_NAME \
                                --key-type rsa \
                                --query "KeyMaterial" \
                                --output text > "/tmp/${KEY_NAME}.pem"

                            if [ $? -ne 0 ]; then
                                echo "Error creating key pair"
                                exit 1
                            fi

                        '''
                    }
                }
            }
        }
        stage('Push Key to Vault') {
            environment {
                ARTIFACTORY_CREDS  = credentials('17777470-1281-48fb-8884-ab80b9e069f0')
            }

            steps {
                script {
                    vaultMod = load './cp4saas/lib/vault.groovy'
                    vaultMod.installCLI(ARTIFACTORY_CREDS)

                    B64_KEY = sh(script: ""'cat /tmp/xdr-bootnode.pem | base64 -w 0'"", returnStdout: true)

                    vaultMod.patchAndPut(credentials_location, credID,
                        ["bootnode_privatekey_${AWS_ACCOUNT}_${REGION}=${B64_KEY}"]
                    )
                }

            }
        }
    }
}

**Comment by nvuillam** [comment]
"Give a man a fish and you feed him for a day; teach a man to fish and you feed him for a lifetime.” 😇

To solve your first case, I opened the file with vscode, and a bracket was highlighted.

I removed big blocks , waited each time for vscode-groovy-lint to run, until I found the block that caused the error.

Then I removed smaller blocks, then line by line until I identified the line messing with the linter.

Then by looking at this line, I saw the ,, :)
